feat: update media form for gallery photos

flagrede opened the pull request:
## Description

Closes: 
regen-network/regen-registry#1615
regen-network/regen-registry#1672

- migrate MediaFrom to `react-hook-form`
- migrate `ImageDrop` to `react-hook-form`
- add gallery photo fields
- update metadata JSON-LD types and wiring

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] provided a link to the relevant issue or specification
- [ ] provided instructions on how to test
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### How to test

1. using the shared account go to
2. https://deploy-preview-1884--regen-registry.netlify.app/project-pages/C45-010/edit/media
3. https://deploy-preview-1884--regen-registry.netlify.app/project-pages/618e1012-e433-11ed-999c-0267c2be097b/media

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

flagrede commented:
@erikalogie You can test the new form media here:
- https://deploy-preview-1884--regen-registry.netlify.app/project-pages/C45-010/edit/media
- https://deploy-preview-1884--regen-registry.netlify.app/project-pages/618e1012-e433-11ed-999c-0267c2be097b/media

flagrede commented on web-registry/src/pages/ProjectReview/ProjectReview.tsx:
```diff
@@ -157,10 +157,10 @@ export const ProjectReview: React.FC<React.PropsWithChildren<unknown>> = () => {
-          photo => photo && <Photo src={photo} />,
+          photo => photo && <Photo src={photo['schema:url']} />,
         )}
         {/* TODO: display story image or video https://github.com/regen-network/regen-registry/issues/1615 */}
```
Well, this was not part of the title issue so I assume this would be done separately (I did not take it into account for the estimation too). The PR already is already quite big anyways so might be worth doing separately.

blushi replied:
as discussed this will be addressed in a separate issue

erikalogie commented:
@flagrede 

The error for "You must add at least 2 photos to the photo gallery." should only appear if the user adds just one photo. Otherwise it is acceptable to have no gallery photos. https://monosnap.com/file/HQDc4uTaKZaiNJ53xC1y5K9V40262b

Also, I can't proceed to the next step in the create form to save the media.

I also can't save the photos in the edit form. As soon as I click save, then navigate to another tab, and come back to the media form, the photos I uploaded have disappeared.

I tried to create a new project to see if the media upload works with a new project but the create project button is not working. I get a red banner that says "error creating project".

blushi commented:
> I tried to create a new project to see if the media upload works with a new project but the create project button is not working. I get a red banner that says "error creating project".

@erikalogie I've come across this once yesterday, seems like a bug we should treat separately, could you file an issue for that please?
Reloading the page helped in my case.

flagrede commented:
@erikalogie I fixed the error message to make it appears only when the gallery has one image.

Regarding the other issues, I was unable to reproduce them.
Here's a project that I have just updated in the edit flow (I tried going back and forth with the tabs after saving too): https://deploy-preview-1884--regen-registry.netlify.app/project-pages/C45-010/edit/media
Here's a draft project that I have just updated in the create flow (I were able to save and go to the next step): 
https://deploy-preview-1884--regen-registry.netlify.app/project-pages/2b26dc3e-e4e2-11ed-86b9-0267c2be097b/media

Could you share the projects that you were having an issue with, maybe there're some specific issues with some projects?

blushi commented on web-components/src/components/inputs/new/ImageDrop/ImageDrop.tsx:
```diff
@@ -0,0 +1,184 @@
+import React, { forwardRef, useState } from 'react';
```
is there a way with react-hook-form to set the field as dirty on change from the input component itself? just like it was the case for the formik based component using `setFieldTouched`
this way we don't have to handle this logic at the form level

flagrede replied:
I'm not sure about that, it might not be possible because we're not supposed to change this manually. The problem we have with this form is that we need to add an empty input field at the initialization which makes the form dirty by default. That's why we can't rely on the `isDirty` for this form specifically but for other forms that don't have this constraint it should be fine to rely on `isDirty` only and we shouldn't need to modify it.

blushi commented on web-components/src/components/inputs/new/ImageField/ImageField.tsx:
```diff
@@ -16,14 +16,15 @@ import { useImageFieldStyles } from './ImageField.styles';
   fixedCrop?: Partial<Crop>;
   children: ReactNode;
-  setValue: (value: string) => void;
+  setValue: (value: any) => void;
```
why do we need to switch to `any`?

flagrede replied:
This function used to receive different types of values, I change it back so the value is always a string but forget to change the type too, will update.

blushi commented on web-registry/src/components/organisms/MediaForm/MediaForm.stories.tsx:
```diff
@@ -1,12 +1,13 @@
-const initialValues: MediaFormValues = {
-  'schema:creditText': '',
+const initialValues: MediaFormSchemaType = {
+  'regen:previewPhoto': { 'schema:url': '', 'schema:creditText': '' },
```
could this be just undefined initially?

flagrede replied:
Yes, it can.

blushi commented on web-registry/src/components/organisms/MediaForm/MediaForm.tsx:
```diff
@@ -1,92 +1,105 @@
+      <Form
+        form={form}
+        onSubmit={async data => {
+          const hasError = validateMediaFormForm({
```
```suggestion
          const hasError = validateMediaForm({
```

why do we need the extra validation step btw? this means we need to have the validation logic in two formats: declarative form with zod + validate function
isn't there a way to perform the validation based on zod schema before submitting? running validation on submit is the default behavior for formik based forms for instance: https://formik.org/docs/guides/validation#when-does-validation-run

flagrede replied:
We needed the extra validation at the submit step here because of the empty input field that messes with the validation count. I'll see if I can merge them into refine rules.

blushi reviewed (commented):
thanks for the huge amount of work on that, that was a complex story!
I've reviewed about 2/3 of the code diff and haven't tested yet but here are some first comments.

erikalogie commented:
@flagrede here is screengrab of the issue in the edit flow:

https://user-images.githubusercontent.com/8460237/234860915-9d6bcd82-288e-468d-afc3-b315c124a63c.mov

erikalogie commented:
> > I tried to create a new project to see if the media upload works with a new project but the create project button is not working. I get a red banner that says "error creating project".
> 
> @erikalogie I've come across this once yesterday, seems like a bug we should treat separately, could you file an issue for that please? Reloading the page helped in my case.

Ok, I've opened an issue for this: https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/gh/regen-network/regen-registry/1674

erikalogie commented:
@flagrede I'm experiencing a new issue where the image uploads as a blank image.

I sent you a video on slack because it was too big to upload here

blushi commented:
> Ok, I've opened an issue for this: https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/gh/regen-network/regen-registry/1674
> 
> It did not resolve on reload for me but it did resolve if I log out and log back in.

cool thanks, this is good to know to solve the issue

flagrede commented:
@erikalogie issue we ran into last Thursday seems to be the same as the one for [project creation](https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/gh/regen-network/regen-registry/1674).
Is there anything else to handle for the QA on this story?

erikalogie commented:
> @erikalogie issue we ran into last Thursday seems to be the same as the one for [project creation](https://app.zenhub.com/workspaces/regen-registry-5f8998bec8958d000f4609e2/issues/gh/regen-network/regen-registry/1674). Is there anything else to handle for the QA on this story?

@flagrede It seems that there is still the weird issue with the cropping of the image in the gallery on the page. 
https://monosnap.com/file/WUXrrvvpBSaugf02DavdsVajIC4PXC

Also on mobile the image dialog should have some padding on the sides rather than having the fields extend all the way to the edges. https://monosnap.com/file/xAMOlQvPoYRq69GKavZx4inmAldEHP

flagrede commented:
@erikalogie I've added the mobile padding and also found the issue about the cropping.

erikalogie commented:
@flagrede looks good!

blushi reviewed (approved):
tACK, nice work!

blushi commented on web-registry/src/components/organisms/MediaForm/MediaFormSimple.tsx:
```diff
@@ -20,73 +35,198 @@ export interface MediaErrorsSimple extends MediaBaseErrors {
+    apiServerUrl: apiUri,
+    projectPath,
+  });
+  const handleDelete = gethandleDelete({
```
```suggestion
  const handleDelete = getHandleDelete({
```
